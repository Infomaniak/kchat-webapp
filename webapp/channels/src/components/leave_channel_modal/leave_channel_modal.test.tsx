// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {shallow} from 'enzyme';
import React from 'react';

import type {ChannelType} from '@mattermost/types/channels';

import LeaveChannelModal from 'components/leave_channel_modal/leave_channel_modal';

describe('components/LeaveChannelModal', () => {
    test('should match snapshot, init', () => {
        const props = {
            channel: {
                id: 'channel7',
                name: 'test-channel-7',
                display_name: 'Test Channel 1',
                type: ('P' as ChannelType),
                team_id: 'team-1',
                header: '',
                purpose: '',
                creator_id: '',
                scheme_id: '',
                group_constrained: false,
                create_at: 0,
                update_at: 0,
                delete_at: 0,
                last_post_at: 0,
                last_root_post_at: 0,
            },
            actions: {
                leaveChannel: jest.fn().mockImplementation(() => {
                    const error = {
                        message: 'error leaving channel',
                    };

                    return Promise.resolve({error});
                }),
                getChannelMemberAction: jest.fn(),
                getChannelStats: jest.fn(),
                getChannelMember: jest.fn(),
                updateChannelMemberSchemeRoles: jest.fn().mockResolvedValue({data: true}),
                loadStatusesForProfilesList: jest.fn(),
                addUsersToChannel: jest.fn().mockResolvedValue({data: true}),
                searchProfiles: jest.fn().mockResolvedValue({data: true}),
                getProfilesNotInChannel: jest.fn().mockResolvedValue({data: true}),
                searchAssociatedGroupsForReference: jest.fn().mockResolvedValue({data: true}),
                getTeamStats: jest.fn(),
                closeModal: jest.fn(),
                getProfilesInChannel: jest.fn().mockResolvedValue({data: true}),
                getTeamMembersByIds: jest.fn().mockResolvedValue({data: true}),
                loadProfilesAndReloadChannelMembers: jest.fn(),
            },
            currentUser: {user_id: 'user-1'},
            onExited: jest.fn(),
            callback: jest.fn(),
            profilesInCurrentChannel: [{id: 'user1'}],
            isInvite: true,
        };

        const wrapper = shallow<typeof LeaveChannelModal>(
            <LeaveChannelModal
                {...props}
            />,
        );

        expect(wrapper).toMatchSnapshot();
    });

    test('should fail to leave channel', () => {
        const props = {
            channel: {
                id: 'channel7',
                name: 'test-channel-7',
                display_name: 'Test Channel 1',
                type: ('P' as ChannelType),
                team_id: 'team-1',
                header: '',
                purpose: '',
                creator_id: '',
                scheme_id: '',
                group_constrained: false,
                create_at: 0,
                update_at: 0,
                delete_at: 0,
                last_post_at: 0,
                last_root_post_at: 0,
            },
            actions: {
                leaveChannel: jest.fn().mockImplementation(() => {
                    const error = {
                        message: 'error leaving channel',
                    };

                    return Promise.resolve({error});
                }),
                getChannelMemberAction: jest.fn(),
                getChannelStats: jest.fn(),
                getChannelMember: jest.fn(),
                updateChannelMemberSchemeRoles: jest.fn().mockResolvedValue({data: true}),
                loadStatusesForProfilesList: jest.fn(),
                addUsersToChannel: jest.fn().mockResolvedValue({data: true}),
                searchProfiles: jest.fn().mockResolvedValue({data: true}),
                getProfilesNotInChannel: jest.fn().mockResolvedValue({data: true}),
                searchAssociatedGroupsForReference: jest.fn().mockResolvedValue({data: true}),
                getTeamStats: jest.fn(),
                closeModal: jest.fn(),
                getProfilesInChannel: jest.fn().mockResolvedValue({data: true}),
                getTeamMembersByIds: jest.fn().mockResolvedValue({data: true}),
                loadProfilesAndReloadChannelMembers: jest.fn(),
            },
            currentUser: {user_id: 'user-1'},
            onExited: jest.fn(),
            callback: jest.fn(),
            profilesInCurrentChannel: [{id: 'user1'}],
            isInvite: true,
        };
        const wrapper = shallow<typeof LeaveChannelModal>(
            <LeaveChannelModal
                {...props}
            />,
        );

        const button = wrapper.find('[data-testid="confirmModalButton"]');
        button.simulate('click');

        expect(props.actions.leaveChannel).toHaveBeenCalledTimes(1);
        expect(props.callback).toHaveBeenCalledTimes(0);
        expect(wrapper.find('.leave-button').exists()).toBe(true);
    });

    test(('should show modal if there is one admin in the channel'), async () => {
        const props = {
            channel: {
                id: 'channel1',
                name: 'channel1',
                display_name: 'Test Channel 1',
                type: ('P' as ChannelType),
                team_id: 'team-1',
                header: '',
                purpose: '',
                creator_id: '',
                scheme_id: '',
                group_constrained: false,
                create_at: 0,
                update_at: 0,
                delete_at: 0,
                last_post_at: 0,
                last_root_post_at: 0,
            },
            actions: {
                getChannelMemberAction: jest.fn(),
                loadProfilesAndReloadChannelMembers: jest.fn(),
                leaveChannel: jest.fn(),
                getChannelStats: jest.fn(),
                getChannelMember: jest.fn(),
                updateChannelMemberSchemeRoles: jest.fn().mockResolvedValue({data: true}),
                loadStatusesForProfilesList: jest.fn(),
                addUsersToChannel: jest.fn().mockResolvedValue({data: true}),
                searchProfiles: jest.fn().mockResolvedValue({data: true}),
                getProfilesNotInChannel: jest.fn().mockResolvedValue({data: true}),
                searchAssociatedGroupsForReference: jest.fn().mockResolvedValue({data: true}),
                getTeamStats: jest.fn(),
                closeModal: jest.fn(),
                getProfilesInChannel: jest.fn().mockResolvedValue({data: true}),
                getTeamMembersByIds: jest.fn().mockResolvedValue({data: true}),
            },
            currentUser: {user_id: 'user-1'},
            onExited: jest.fn(),
            callback: jest.fn(),
            currentUserIsChannelAdmin: true,
            profilesInCurrentChannel: [{id: 'user1'}, {id: 'user-2'}],
            isInvite: true,
            currentMemberIsChannelAdmin: true,
        };

        const wrapper = shallow<typeof LeaveChannelModal>(
            <LeaveChannelModal
                {...props}
            />,
        );

        expect(wrapper.find('.test-channel-1-download').exists()).toBe(true);
    });
});
