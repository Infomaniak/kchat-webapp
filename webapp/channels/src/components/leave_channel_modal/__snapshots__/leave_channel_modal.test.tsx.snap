// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/LeaveChannelModal should match snapshot, init 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="modal-open"
    style="padding-right: 0px;"
  >
    <div
      aria-hidden="true"
    />
    <div
      role="dialog"
    >
      <div
        class="fade modal-backdrop in"
      />
      <div
        aria-labelledby="channelInviteModalLabel"
        class="fade in modal"
        id="addUsersToChannelModal"
        role="dialog"
        style="display: block; padding-right: 0px;"
        tabindex="-1"
      >
        <div
          class="a11y__modal channel-invite modal-dialog"
        >
          <div
            class="modal-content"
            role="document"
          >
            <div
              class="modal-header"
              id="channelInviteModalLabel"
            >
              <button
                aria-label="Close"
                class="close"
                type="button"
              >
                <span
                  aria-hidden="true"
                >
                  Ã—
                </span>
                <span
                  class="sr-only"
                >
                  Close
                </span>
              </button>
              <h1
                class="modal-title"
                id="deletePostModalLabel"
              >
                Leave Private Channel 
                <b>
                  Test Channel 1
                </b>
              </h1>
            </div>
            <div
              class="overflow--visible modal-body"
              role="application"
            >
              <div
                class="channel-invite__content"
              >
                <div>
                  <div
                    class="alert alert-with-icon-leave alert-grey"
                  >
                    <i
                      class="icon-information-outline"
                    />
                    Only a channel administrator can invite you to join this channel again.
                  </div>
                  <div
                    class="alert-message"
                  >
                    Do you really want to leave the 
                    <b>
                      Test Channel 1
                    </b>
                     channel?
                  </div>
                </div>
              </div>
            </div>
            <div
              class="modal-footer"
            >
              <button
                class="btn btn-tertiary"
                id="cancelModalButton"
                type="button"
              >
                Cancel
              </button>
              <div
                class="leave-button"
              >
                <button
                  class="btn btn-danger"
                  data-testid="confirmModalButton"
                  id="confirmModalButton"
                  type="button"
                >
                  Leave
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "replaceStoreState": [Function],
  "rerender": [Function],
  "unmount": [Function],
  "updateStoreState": [Function],
}
`;
